pkg/resourcevalidator/webhook.go

```diff
@@ -0,0 +1,241 @@
+	"fmt"
+	"net/http"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	vkv1alpha1 "github.com/liqotech/liqo/apis/virtualkubelet/v1alpha1"
+	"github.com/liqotech/liqo/pkg/consts"
+	liqogetters "github.com/liqotech/liqo/pkg/utils/getters"
+	liqolabels "github.com/liqotech/liqo/pkg/utils/labels"
+	pod "github.com/liqotech/liqo/pkg/utils/pod"
+	"github.com/liqotech/liqo/pkg/virtualKubelet/forge"
+)
+
+// cluster-role
+// +kubebuilder:rbac:groups=virtualkubelet.liqo.io,resources=shadowpods,verbs=get;list;watch
+// +kubebuilder:rbac:groups=sharing.liqo.io,resources=resourceoffers,verbs=get;list;watch
+// +kubebuilder:rbac:groups=core,resources=namespaces,verbs=get;list;watch
+
+// Manifests
+// flag on controller manager
+
+// ShadowPodValidator is the handler used by the Validating Webhook to validate shadow pods.
+type ShadowPodValidator struct {
+	Client                   client.Client
+	PeeringCache             *peeringCache
+	decoder                  *admission.Decoder
+	enableResourceValidation bool
+}
+
+// NewShadowPodValidator creates a new shadow pod validator.
+func NewShadowPodValidator(c client.Client, enableResourceValidation bool) *ShadowPodValidator {
+	return &ShadowPodValidator{
+		Client:                   c,
+		PeeringCache:             &peeringCache{ready: false},
+		enableResourceValidation: enableResourceValidation,
+	}
+}
+
+// Handle is the funcion in charge of handling the request.
+// nolint:gocritic // the signature of this method is imposed by controller runtime.
+func (spv *ShadowPodValidator) Handle(ctx context.Context, req admission.Request) admission.Response {
+	webhooklog.Info(fmt.Sprintf("\t\tOperation: %s", req.Operation))
+
+	if !spv.PeeringCache.ready {
+		webhooklog.Info("PeeringCache not ready")
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("PeeringCache not ready"))
+	}
+
+	var obj runtime.RawExtension
+
+	switch req.Operation {
+	case admissionv1.Create:
+		obj = req.Object
+	case admissionv1.Delete:
+		obj = req.OldObject
+	case admissionv1.Update:
+		return spv.HandleUpdate(req)
+	default:
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("unsupported operation %s", req.Operation))
+	}
+
+	// Decode the shadow pod
+	shadowpod, err := spv.DecodeShadowPod(obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Check existence and get shadow pod origin Cluster ID label
+	spClusterID, found := shadowpod.Labels[forge.LiqoOriginClusterIDKey]
+	if !found {
+		return admission.Denied("missing origin Cluster ID label")
+	}
+
+	// Get ShadowPod Namespace
+	namespace := &corev1.Namespace{}
+	if err := spv.Client.Get(ctx, client.ObjectKey{Name: shadowpod.GetNamespace()}, namespace); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
```

Comment:
This should be an InternalServerError, and not a client error. The error message shall be enriched before being returned to the caller.